src/lib/patterns.js

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const glob = require('fast-glob');
+const fs = require('fs');
+const path = require('path');
+const matter = require('gray-matter');
+const md = require('markdown-it')();
+
+const stripDot = /^\./;
+const basePath = 'src/site/content/en/patterns';
+const files = glob.sync(path.join(basePath, '**', 'index.md'));
+
+const ids = new Set();
+
+const allPatterns = files
+  .map((file) => {
+    const id = path.relative(basePath, path.dirname(file));
```

Comment:
This basically seems like we're checking that no two files have the same path. Is that right? Is this helpful? I feel like it'll never fire for any real file-system.

Reply:
Right